internal/controller/config/config.go

```diff
@@ -0,0 +1,98 @@
+package config
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type ConfigManager struct {
+	config *GlobalConfig
+	lock   *sync.RWMutex
+}
+
+var instance *ConfigManager
+var once sync.Once
+
+// GetConfigManagerInstance  returns a singleton config manager throughout the application
+func GetConfigManagerInstance() *ConfigManager {
+	once.Do(func() {
+		instance = &ConfigManager{
+			config: &GlobalConfig{},
+			lock:   new(sync.RWMutex),
+		}
+	})
+	return instance
+}
+
+// GlobalConfig is the configuration for the controllers, it is
+// supposed to be populated from the configmap attached to the
+// controller manager.
+type GlobalConfig struct {
+	ClusterName     string `json:"clusterName"`
+	TimeIntervalSec uint   `json:"timeIntervalSec"`
+	// RepoCredentials maps each Git Repository Path prefix to the corresponding credentials that are needed for it
+	RepoCredentials map[string]*GitCredential `json:"repoCredentials"`
+}
+
+type GitCredential struct {
+	HTTPCredential *HTTPCredential `json:"HTTPCredential"`
+	SSHCredential  *SSHCredential  `json:"SSHCredential"`
+	TLS            *TLS            `json:"TLS"`
+}
+
+type HTTPCredential struct {
+	Username string            `json:"username"`
+	Password SecretKeySelector `json:"password"`
+}
+
+type SSHCredential struct {
+	SSHKey SecretKeySelector `json:"SSHKey" yaml:"SSHKey" `
+}
+
+type TLS struct {
+	InsecureSkipVerify bool              `json:"insecureSkipVerify"`
+	CACertSecret       SecretKeySelector `json:"CACertSecret"`
+	CertSecret         SecretKeySelector `json:"CertSecret"`
+	KeySecret          SecretKeySelector `json:"keySecret"`
+}
+
+type SecretKeySelector struct {
+	corev1.LocalObjectReference `mapstructure:",squash"` // for viper to correctly parse the config
+	Key                         string                   `json:"key" `
+	Optional                    *bool                    `json:"optional,omitempty" `
+}
+
+func (cm *ConfigManager) GetConfig() *GlobalConfig {
+	cm.lock.Lock()
+	defer cm.lock.Unlock()
```

Comment:
```suggestion
	cm.lock.RLock()
	defer cm.lock.RUnlock()
```

Reply:
(this is a reader.....using this function allows multiple readers to read at the same time, which is deemed safe)